experiment/measurer/measure_manager.py

```diff
@@ -396,6 +397,82 @@ def generate_profdata(self, cycle: int):
+        time_stamp = cycle
+
+        # Create function dataframe and load previous segment dataframe.
+        function_coverage_state_file = measure_worker.StateFile(
+            measure_worker.FUNCTION_COVERAGE_STATE_NAME, self.state_dir, cycle)
+        function_coverage = pandas.DataFrame(columns=[
+            'benchmark', 'fuzzer', 'trial', 'time', 'function', 'hits'
+        ])
+
+        segment_coverage_state_file = measure_worker.StateFile(
+            measure_worker.SEGMENT_COVERAGE_STATE_NAME, self.state_dir, cycle)
+        if not segment_coverage_state_file.get_previous():
+            segment_coverage = pandas.DataFrame(columns=[
+                'benchmark', 'fuzzer', 'trial', 'time', 'file', 'line', 'column'
+            ])
+        else:
+            segment_coverage = pandas.from_dict(
+                segment_coverage_state_file.get_previous(), orient='table')
+
+        # Extract coverage information
+        try:
+            coverage_info = coverage_utils.get_coverage_infomation(
+                self.cov_summary_file)
+            for function_data in coverage_info['data'][0]['functions']:
+                entry = pandas.Series([
+                    self.benchmark, self.fuzzer, self.trial_num,
+                    function_data['name'], function_data['count'], time_stamp
+                ],
+                                      index=function_coverage.columns)
+                function_coverage = function_coverage.append(entry,
+                                                             ignore_index=True)
+
+            for file in coverage_info['data'][0]['files']:
+                for segment in file['segments']:
+                    if segment[2] != 0:  # Segment hits.
+                        entry = pandas.Series(
+                            [
+                                self.benchmark,
+                                self.fuzzer,
+                                self.trial_num,
+                                file['filename'],
+                                segment[0],  # Segment line.
+                                segment[1],  # Segment column.
+                                time_stamp
+                            ],
+                            index=segment_coverage.columns)
+                        segment_coverage = segment_coverage.append(
+                            entry, ignore_index=True)
+
+        except (ValueError, KeyError, IndexError):
+            coverage_utils.logger.error(
+                'Failed to extract trial-specific segment and function '
+                'information from coverage summary.')
+
+        # Drop segments that have previously been discovered.
+        try:
+            segment_coverage = segment_coverage.sort_values(by=['time'])
+            segment_coverage = segment_coverage.drop_duplicates(
+                subset=segment_coverage.columns.difference(['time']),
+                keep='first')
```

Comment:
I think instead of adding all the segments and then removing the redundant ones would be memory expensive (we will add about 30K segment entries and then remove duplicates). Rather we can add it only if it is not in the data frame already. 

While adding the entry we can check if it is already present like this

`if [segment[0], segment[1]] not in [[segment_coverage['line'][i], segment_coverage['column'][i]] for i in 
     segment_coverage.index]:`
       `segment_coverage = segment_coverage.append(entry, ignore_index=True)`

Reply:
Or actually we should pre-compute all the unique segments available with us already and then check for each entry if it is redundant.

Reply:
I think this can be further simplified for added performance. Let me check.

 Maybe this can. be re-written as :
 `recorded_segments = [list(x) for x in zip(segment_coverage['line'].to_list(), segment_coverage['column'].to_list(),)]`